mysql service doesn't work with package=mysql57

## Issue description

If `services.mysql` is configured to use `pkgs.mysql57`, the `mysql` service fails to start, the logs contain:

```
Nov 04 07:56:12 misctests-rigel systemd[1]: Starting MySQL Server...
Nov 04 07:56:12 misctests-rigel kdg5ciqxdbc1gh8bk3brmy9j641xp8xx-unit-script-mysql-pre-start[11246]: mysql_install_db: [ERROR] unknown variable 'pid-file=/run/mysqld/mysqld.pid'
Nov 04 07:56:12 misctests-rigel kdg5ciqxdbc1gh8bk3brmy9j641xp8xx-unit-script-mysql-pre-start[11246]: 2018-11-04 07:56:12 [ERROR]   Unrecognized options
```

### Steps to reproduce

Place the following in NixOS configuration:

```
services.mysql.enable = true;
services.mysql.package = pkgs.mysql57;
```

Ensure that `/var/lib/mysql` does not exist.

Run `nixos-rebuild switch`.

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.76, NixOS, 18.09.892.c06f5302f63 (Jellyfish)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.1.1`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

--- a/nixos/modules/services/databases/mysql.nix
+++ b/nixos/modules/services/databases/mysql.nix
@@ -12,12 +12,22 @@ let
     let
       pName = _p: (builtins.parseDrvName (_p.name)).name;
     in pName mysql == pName pkgs.mariadb;
+  isMysqlAtLeast57 =
+    let
+      pName = _p: (builtins.parseDrvName (_p.name)).name;
+    in (pName mysql == pName pkgs.mysql57)
+       && ((builtins.compareVersions mysql.version "5.7") >= 0);
 
   pidFile = "${cfg.pidDir}/mysqld.pid";
 
+  mysqldAndInstallOptions =
+    "--user=${cfg.user} --datadir=${cfg.dataDir} --basedir=${mysql}";
   mysqldOptions =
-    "--user=${cfg.user} --datadir=${cfg.dataDir} --basedir=${mysql} " +
-    "--pid-file=${pidFile}";
+    "${mysqldAndInstallOptions} --pid-file=${pidFile}";
+  # For MySQL 5.7+, --insecure creates the root user without password
+  # (earlier versions and MariaDB do this by default).
+  installOptions =
+    "${mysqldAndInstallOptions} ${lib.optionalString isMysqlAtLeast57 "--insecure"}";
 
   myCnf = pkgs.writeText "my.cnf"
   ''
@@ -252,7 +262,7 @@ in
             if ! test -e ${cfg.dataDir}/mysql; then
                 mkdir -m 0700 -p ${cfg.dataDir}
                 chown -R ${cfg.user} ${cfg.dataDir}
-                ${mysql}/bin/mysql_install_db ${mysqldOptions}
+                ${mysql}/bin/mysql_install_db ${installOptions}
                 touch /tmp/mysql_init
             fi
 

